scenario_player/tasks/base.py

```diff
@@ -108,8 +110,28 @@ def __call__(self, *args, **kwargs):
         self._runner.running_task_count += 1
         self._start_time = time.monotonic()
         try:
-            return_val = self._run(*args, **kwargs)
-            gevent.sleep(self.SYNCHRONIZATION_TIME_SECONDS)
+            # Zero means no timeout is desired
+            timeout_s = self._config.get("timeout", self.DEFAULT_TIMEOUT)
+            if timeout_s > 0:
+                exception: Optional[Exception] = None
+                try:
+                    with Timeout(self._config.get("timeout", self.DEFAULT_TIMEOUT)):
+                        return_val = None
+                        while True:
+                            try:
+                                return_val = self._run(*args, **kwargs)
+                            except ScenarioAssertionError as ex:
+                                exception = ex
+
+                            if return_val:
+                                return return_val
+
+                            sleep(1)
+                except Timeout:
+                    if exception:
+                        raise exception
```

Comment:
It would probably be good to add a bit of logging here (e.g. retry count, maybe remaining timeout...)